File: sdk/python/examples/live_run.py
"""Connect the Python SDK using a one-use credential from an anonymous stdin pipe."""

from __future__ import annotations

import asyncio
import os
import stat
import sys
from pathlib import Path

from colossus.api.v1alpha1 import system_pb2
from durable_run import run_prompt

from colossus_sdk import ColossusClient, EndpointDescriptor, StaticBearerCredential

_MAX_CREDENTIAL_BYTES = 761


def read_pipe_credential() -> StaticBearerCredential:
    """Read one bounded bearer only when stdin is a pipe or local socket."""

    mode = os.fstat(sys.stdin.fileno()).st_mode
    if not (stat.S_ISFIFO(mode) or stat.S_ISSOCK(mode)):
        raise RuntimeError("the live SDK credential must arrive through an anonymous pipe")
    raw = sys.stdin.buffer.read(_MAX_CREDENTIAL_BYTES + 1)
    if not raw or len(raw) > _MAX_CREDENTIAL_BYTES:
        raise RuntimeError("the live SDK credential is invalid")
    try:
        token = raw.decode("ascii")
    except UnicodeDecodeError as error:
        raise RuntimeError("the live SDK credential is invalid") from error
    return StaticBearerCredential(token)


async def main() -> None:
    if len(sys.argv) != 6:
        raise RuntimeError(
            "usage: live_run.py DESCRIPTOR CERTIFICATE INSTANCE_ID CERTIFICATE_SHA256 PROMPT"
        )
    descriptor_path, certificate_path, instance_id, certificate_sha256, prompt = sys.argv[1:]
    descriptor = EndpointDescriptor.from_json(Path(descriptor_path).read_text(encoding="utf-8"))
    certificate = Path(certificate_path).read_bytes()
    credential = read_pipe_credential()
    client = await ColossusClient.connect(
        descriptor,
        certificate,
        instance_id,
        certificate_sha256,
        system_pb2.DEPLOYMENT_MODE_SHARED_DAEMON,
        credential,
    )
    try:
        result = await run_prompt(client, prompt)
        print(result.output)
    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/typescript/examples/live-run.ts
/**
 * Connect the TypeScript SDK using a one-use credential from an anonymous stdin pipe.
 */

import { fstatSync, readFileSync } from "node:fs";

import {
  StaticBearerCredential,
  createSecureGrpcClient,
  parseEndpointDescriptor,
} from "../src/index.js";
import { AgentRunServiceClient } from "../src/gen/colossus/api/v1alpha1/agent_run.js";
import { DeploymentMode } from "../src/gen/colossus/api/v1alpha1/system.js";

import { runPrompt } from "./durable-run.js";

const MAX_CREDENTIAL_BYTES = 761;

function readPipeCredential(): StaticBearerCredential {
  const stdin = fstatSync(0);
  if (!stdin.isFIFO() && !stdin.isSocket()) {
    throw new Error(
      "the live SDK credential must arrive through an anonymous pipe",
    );
  }
  const token = readFileSync(0);
  if (token.length === 0 || token.length > MAX_CREDENTIAL_BYTES) {
    throw new Error("the live SDK credential is invalid");
  }
  return new StaticBearerCredential(token.toString("ascii"));
}

async function main(): Promise<void> {
  const [descriptorPath, certificatePath, instanceId, certificateSha256, prompt] =
    process.argv.slice(2);
  if (
    descriptorPath === undefined ||
    certificatePath === undefined ||
    instanceId === undefined ||
    certificateSha256 === undefined ||
    prompt === undefined
  ) {
    throw new Error(
      "usage: live-run DESCRIPTOR CERTIFICATE INSTANCE_ID CERTIFICATE_SHA256 PROMPT",
    );
  }
  const descriptor = parseEndpointDescriptor(
    readFileSync(descriptorPath, "utf8"),
  );
  const certificate = readFileSync(certificatePath);
  const credential = readPipeCredential();
  const client = await createSecureGrpcClient(
    AgentRunServiceClient,
    descriptor,
    certificate,
    instanceId,
    certificateSha256,
    DeploymentMode.DEPLOYMENT_MODE_SHARED_DAEMON,
    credential,
  );
  try {
    const result = await runPrompt(client, prompt);
    process.stdout.write(`${result.output}\n`);
  } finally {
    client.close();
  }
}

await main();
